Cards/CardBase.gd

```diff
@@ -43,19 +43,44 @@ func _apply_all_effects(target: Entity) -> void:
 	for effect_data: EffectData in card_effects_data:
 		effect_data.apply_effect_data(target)
 
+func _apply_effect(target : Entity, effect: EffectData):
+	effect.apply_effect_data(target)
 
 func can_play_card(caster: Entity, target: Entity) -> bool:
 	return caster.get_party_component().can_play_on_entity(application_type, target)
 
 
-func on_card_play(caster: Entity, target: Entity) -> void:
-	_apply_all_effects(target)
-	_deal_damage(caster, target)
-	_apply_status(caster, target)
-	_draw_cards()
-	_discard_random_cards()
+func on_card_play(caster: Entity, targets: Array[Entity]) -> void:
+	#Split up targeted attacks and all attacks
+	var target_effects: Array[EffectData] = []
+	var all_effects: Array[EffectData] = []
+	
+	for effect_data: EffectData in card_effects_data:
+		if(effect_data.target_type == Enums.TargetType.SINGLE_TARGET):
+			target_effects.push_front(effect_data)
+		else :
+			all_effects.push_front( effect_data)
+	#Apply single target
+	for entity : Entity in targets:
+		for effect : EffectData in target_effects:
+			_apply_effect(entity, effect)
+	#Get every unit that is to be affected by card
+	var all_target : Array[Entity]
+	
+	match application_type:
+		Enums.ApplicationType.ALL:
+			all_target = CardManager.card_container.battler_refrence._enemy_list
+			all_target += [PlayerManager.player]
+		Enums.ApplicationType.ENEMY_ONLY:
+			all_target = CardManager.card_container.battler_refrence._enemy_list
+		Enums.ApplicationType.FRIENDLY_ONLY:
+			all_target = [PlayerManager.player]
+	#apply effect to every target
+	for entity : Entity in all_target:
+		for effect : EffectData in all_effects:
+			_apply_effect(entity, effect)
+	
```

Comment:
I think it will probably be better to move this logic into the Battler, then we could remove the `battler_reference` that we have here. However I believe Turtyo has mentioned there will be future changes to how targeting works. So for now this works fine I suppose.

Reply:
I guess, but it feels more intuitve for me that playing cards happens in the card script, but we can see what's better if changes happen

Reply:
I agree that matching the type should probably be moved to the battler and given to the card_play maybe. And yes, I will open an issue for the targeting i'm thinking of to see if we agree.
The problem of having all the logic by inside `on_card_play` is also that we will need to reformat it every time we want to change something about the logic. If instead we use functions that do things, we only need to change the sub-function without needing to change the card play in itself. `on_card_play` should only deal with having to play the card, and not knowing how to build the list of effects or knowing how to build the list of targets. Those things should be dealt in their respective functions, and `on_card_play` should only call them.

Reply:
That's a good point, I'll at least move some of the logic out to other functions

Reply:
Could you give us a little summary of what you moved out ? I don't remember how it was looking at that point

Reply:
I moved out the filtering of single and all target attacks into get_effects_targeting_single and get_effects_target_all respectively, as well as how the function got the array of every target into get_all_targets in the battle manager

Reply:
alright seems nice enough
changes will be made probably due to the targeting PR, but this looks okay
@cheesefrycook feel free to resolve if you feel like it's ok